zebrad/src/components/miner.rs

```diff
@@ -0,0 +1,527 @@
+        > + Clone
+        + Send
+        + Sync
+        + 'static,
+    Mempool::Future: Send,
+    State: Service<
+            zebra_state::ReadRequest,
+            Response = zebra_state::ReadResponse,
+            Error = zebra_state::BoxError,
+        > + Clone
+        + Send
+        + Sync
+        + 'static,
+    <State as Service<zebra_state::ReadRequest>>::Future: Send,
+    Tip: ChainTip + Clone + Send + Sync + 'static,
+    BlockVerifierRouter: Service<zebra_consensus::Request, Response = block::Hash, Error = zebra_consensus::BoxError>
+        + Clone
+        + Send
+        + Sync
+        + 'static,
+    <BlockVerifierRouter as Service<zebra_consensus::Request>>::Future: Send,
+    SyncStatus: ChainSyncStatus + Clone + Send + Sync + 'static,
+    AddressBook: AddressBookPeers + Clone + Send + Sync + 'static,
+{
+    let configured_threads = config.internal_miner_threads;
+    // If we can't detect the number of cores, use the configured number.
+    let available_threads = available_parallelism()
+        .map(usize::from)
+        .unwrap_or(configured_threads);
+
+    // Use the minimum of the configured and available threads.
+    let solver_count = min(configured_threads, available_threads);
+
+    info!(
+        ?solver_count,
+        "launching mining tasks with parallel solvers"
+    );
+
+    let (template_sender, template_receiver) = watch::channel(None);
+    let template_receiver = WatchReceiver::new(template_receiver);
+
+    // Spawn these tasks, to avoid blocked cooperative futures, and improve shutdown responsiveness.
+    // This is particularly important when there are a large number of solver threads.
+    let mut abort_handles = Vec::new();
+
+    let template_generator = tokio::task::spawn(
+        generate_block_templates(rpc.clone(), template_sender).in_current_span(),
+    );
+    abort_handles.push(template_generator.abort_handle());
+    let template_generator = template_generator.wait_for_panics();
+
+    let mut mining_solvers = FuturesUnordered::new();
+    for solver_id in 0..solver_count {
+        // Assume there are less than 256 cores. If there are more, only run 256 tasks.
+        let solver_id = min(solver_id, usize::from(u8::MAX))
+            .try_into()
+            .expect("just limited to u8::MAX");
+
+        let solver = tokio::task::spawn(
+            run_mining_solver(solver_id, template_receiver.clone(), rpc.clone()).in_current_span(),
+        );
+        abort_handles.push(solver.abort_handle());
+
+        mining_solvers.push(solver.wait_for_panics());
+    }
+
+    // These tasks run forever unless there is a fatal error or shutdown.
+    // When that happens, the first task to error returns, and the other JoinHandle futures are
+    // cancelled.
+    let first_result;
+    select! {
+        result = template_generator => { first_result = result; }
+        result = mining_solvers.try_next() => {
+            first_result = result
+                .transpose()
```

Comment:
Nitpick: Use `futures::StreamExt` instead of `TryStreamExt`

```suggestion
        result = mining_solvers.next() => {
            first_result = result
```

Reply:
Happy to do this, but I'd like to focus on bugs or wider code structure in this review.